test/jdk/java/lang/reflect/records/IsRecordTest.java

```diff
@@ -0,0 +1,196 @@
+
+        List<RecordComponentEntry> rc = null;
+        if (withRecordAttr)
+            rc = List.of(new RecordComponentEntry("x", "I"));
+        String superName = extendsJLR ? "java/lang/Record" : "java/lang/Object";
+        var classBytes = generateClassBytes("C", isFinal, isAbstract, superName, rc);
+        Class<?> cls = ByteCodeLoader.load("C", classBytes);
+        out.println("cls=%s, Record::isAssignable=%s, isRecord=%s"
+                .formatted(cls, Record.class.isAssignableFrom(cls), cls.isRecord()));
+        assertEquals(cls.isRecord(), expectIsRecord);
+        var getRecordComponents = cls.getRecordComponents();
+        assertTrue(expectIsRecord ? getRecordComponents != null : getRecordComponents == null);
+    }
+
+    /**
+     * Tests the valid combinations of i) final/non-final, ii) abstract/non-abstract,
+     * along with the presence or absence of a record attribute, where the class has
+     * a superclass whose superclass is j.l.Record.
+     */
+    @Test(dataProvider = "scenarios")
+    public void testIndirectSubClass(boolean isFinal,
+                                     boolean isAbstract,
+                                     boolean unused1,
+                                     boolean withRecordAttr,
+                                     boolean unused2) throws Exception {
+        out.println("\n--- testIndirectSubClass isFinal=%s, isAbstract=%s withRecordAttr=%s ---"
+                .formatted(isFinal, isAbstract, withRecordAttr));
+
+        List<RecordComponentEntry> rc = null;
+        if (withRecordAttr)
+            rc = List.of(new RecordComponentEntry("x", "I"));
+        var supFooClassBytes = generateClassBytes("SupFoo", false, isAbstract, "java/lang/Record", rc);
+        var subFooClassBytes = generateClassBytes("SubFoo", isFinal, isAbstract, "SupFoo", rc);
+        var allClassBytes = Map.of("SupFoo", supFooClassBytes,
+                                   "SubFoo", subFooClassBytes);
+
+        ClassLoader loader = new ByteCodeLoader(allClassBytes, null);
+        Class<?> supFooCls = loader.loadClass("SupFoo");
+        Class<?> subFooCls = loader.loadClass("SubFoo");
+        for (var cls : List.of(supFooCls, subFooCls))
+            out.println("cls=%s, Record::isAssignable=%s, isRecord=%s"
+                    .formatted(cls, Record.class.isAssignableFrom(cls), cls.isRecord()));
+        assertFalse(supFooCls.isRecord());
+        assertFalse(subFooCls.isRecord());
+        assertEquals(supFooCls.getRecordComponents(), null);
+        assertEquals(subFooCls.getRecordComponents(), null);
+    }
+
+    /** Tests record-ness properties of traditionally compiled classes. */
+    @Test
+    public void testBasicRecords() {
+        out.println("\n--- testBasicRecords ---");
+        record EmptyRecord () { }
+        assertTrue(EmptyRecord.class.isRecord());
+        assertTrue(EmptyRecord.class.getRecordComponents() != null);
```

Comment:
Better to have a more precise check for empty array

Reply:
Good idea - done.